File: problem5.py
def primes():
    yield 2
    ps = [2]
    n = 3
    while True:
        for p in ps:
            if n % p == 0:
                break
            if p**2 >= n:
                ps.append(n)
                yield n
                break
        n += 2

def countFactor(n, p):
    c = 0
    while n % p == 0:
        c += 1
        n //= p
    return c

def maxFactors(n):
    for p in primes():
        if p > n:
            break
        yield p, max(countFactor(m, p) for m in range(1, n+1))

def main():
    prod = 1
    for i, j in maxFactors(20):
        prod *= i ** j
    print(prod)

main()

File: problem5.hs
primes = 2:filter f [3,5..]
    where f n = all ((/= 0) . (mod n)) (takeWhile ((<= n) . (^ 2)) primes)

countFactor n p =
    if n `mod` p == 0 then (1 + countFactor (n `div` p) p) else 0

smallestMultiple n =
        product $ zipWith (^) primesN (map (maxFactors n) primesN)
    where
        -- nまでの素数
        primesN :: [Integer]
        primesN = takeWhile (<n) primes
        -- [1..n] のうち、 p の最大の因数
        maxFactors :: Integer -> Integer -> Integer
        maxFactors n p = maximum [countFactor x p | x <- [1..n]]

main = putStrLn $ show $ smallestMultiple 20
